### PR TITLE
[WIP] Member users with access to specific services should only see accounts subscribed to those services

### PR DESCRIPTION
**Pull request by HemaHG**
closes [THREESCALE-1608](https://issues.redhat.com/browse/THREESCALE-1608)

### CODE REVIEW

File: test/integration/buyers/accounts_controller_test.rb [hunk 1]
```diff
@@ -294,6 +294,37 @@ def setup
...
+      member = FactoryBot.create(:member, account: @provider,
+                                      member_permission_ids: [:partners])
+      member.activate!
+      FactoryBot.create(:member_permission, user: member,
+                                        admin_section: :services,
+                                        service_ids: [service2.id])
+
+
+      login_provider @provider, user: member
+    end
```
Comment (guicassolato) [inline test/integration/buyers/accounts_controller_test.rb:327]: I'd try something like this out:

```suggestion
    test "check if user with access to specific service can see accounts particular to that service" do
      buyer_without_applications = @provider.buyers.first

      accessible_service = FactoryBot.create(:service, account: @provider)
      accessible_buyer = FactoryBot.create(:buyer_account, provider_account: @provider)
      accessible_app_plan = FactoryBot.create(:application_plan, issuer: accessible_service)
      accessible_app = FactoryBot.create(:cinstance, user_account: accessible_buyer, plan: accessible_app_plan, name: 'accessbile app')

      forbidden_service = FactoryBot.create(:service, account: @provider)
      forbidden_buyer = FactoryBot.create(:buyer_account, provider_account: @provider)
      forbidden_app_plan = FactoryBot.create(:application_plan, issuer: forbidden_service)
      forbidden_app = FactoryBot.create(:cinstance, user_account: forbidden_buyer, plan: forbidden_app_plan, name: 'forbidden app')

      # member users with restricted access to services should see accounts subscribed to the service
      member = FactoryBot.create(:member, account: @provider, member_permission_ids: [:partners])
      member.member_permission_service_ids = [accessible_service.id]
      member.activate!

      logout! && login!(@provider, user: member)

      get admin_buyers_accounts_path
      assert_equal 2, assigns(:accounts).size
      assert_includes assigns(:accounts), accessible_buyer
      assert_includes assigns(:accounts), buyer_without_applications
      assert_not_includes assigns(:accounts), forbidden_buyer
    end
  end
```

File: app/models/user.rb [hunk 2]
```diff
@@ -226,6 +226,13 @@ def accessible_cinstances
     end
   end
 
+  def accessible_buyer_accounts
+    buyer_accounts = account.buyer_accounts
+    buyer_accounts_without_applications = buyer_accounts.where.has { not_exists Cinstance.by_account(BabySqueel[:accounts].id) }
+    buyer_accounts.where(id:
+    accessible_cinstances.select(:user_account_id)).or(buyer_accounts_without_applications)
```
Comment (guicassolato) [inline app/models/user.rb:233]:
```suggestion
    buyer_accounts.where(id: accessible_cinstances.select(:user_account_id)).or(buyer_accounts_without_applications)
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 3]
```diff
@@ -163,6 +163,54 @@ def test_show
       end
       assert_response :forbidden
     end
+
+    test 'member users with access to all services should see all buyers accounts' do
+      buyer = FactoryBot.create(:simple_buyer, provider_account: provider)
```
Comment (guicassolato) [inline test/integration/buyers/accounts_controller_test.rb:168]: It looks like we are not using this buyer in the test.

File: test/integration/buyers/accounts_controller_test.rb [hunk 4]
```diff
@@ -163,6 +163,54 @@ def test_show
...
+      accessible_app_plan = FactoryBot.create(:application_plan, issuer: accessible_service)
+      accessible_app = FactoryBot.create(:cinstance, user_account: accessible_buyer, plan: accessible_app_plan, name: 'accessible app')
+
+      another_service = FactoryBot.create(:service, account: @provider)
+      another_buyer = FactoryBot.create(:buyer_account, provider_account: @provider)
+      another_app_plan = FactoryBot.create(:application_plan, issuer: another_service)
+      another_app = FactoryBot.create(:cinstance, user_account: another_buyer, plan: another_app_plan, name: 'another app')
+
+      user.member_permission_ids = [:partners]
+      user.member_permission_service_ids = [accessible_service.id, another_service.id]
```
Comment (guicassolato) [inline test/integration/buyers/accounts_controller_test.rb:182]: This is not "all services". I'd skip this line for this test case.

File: test/integration/buyers/accounts_controller_test.rb [hunk 5]
```diff
@@ -163,6 +163,54 @@ def test_show
...
+      user.save!
+      get admin_buyers_accounts_path
+      assert_equal 3, assigns(:accounts).size
+      assert_includes assigns(:accounts), accessible_buyer
+      assert_includes assigns(:accounts), buyer_without_applications
+      assert_includes assigns(:accounts), another_buyer
+    end
+
+    test 'member users with restricted access to services should see accounts subscribed to the service and accounts with no applications at all' do
+      buyer = FactoryBot.create(:simple_buyer, provider_account: provider)
```
Comment (guicassolato) [inline test/integration/buyers/accounts_controller_test.rb:192]: Same as https://github.com/3scale/porta/pull/2233/files#r515111751

File: test/integration/buyers/accounts_controller_test.rb [hunk 6]
```diff
@@ -294,6 +342,26 @@ def setup
...
+      accessible_service2 = FactoryBot.create(:service, account: @provider)
+      accessible_buyer2 = FactoryBot.create(:buyer_account, provider_account: @provider)
+      accessible_app_plan2 = FactoryBot.create(:application_plan, issuer: accessible_service2)
+      accessible_app2 = FactoryBot.create(:cinstance, user_account: accessible_buyer2, plan: accessible_app_plan2, name: 'accessible2 app')
+
+      get admin_buyers_accounts_path
+      assert_equal 3, assigns(:accounts).size
+      assert_includes assigns(:accounts), accessible_buyer
+      assert_includes assigns(:accounts), buyer_without_applications
+      assert_includes assigns(:accounts), accessible_buyer2
```
Comment (guicassolato) [inline test/integration/buyers/accounts_controller_test.rb:363]:
```suggestion
      assert_same_elements @provider.buyers, assigns(:accounts)
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 7]
```diff
@@ -294,6 +342,26 @@ def setup
       get admin_buyers_accounts_path(id: @buyer.id)
       assert_xpath('//tbody/tr/td[2]/a', @buyer.decorate.admin_user_display_name)
     end
+
+    test "admin user should see all buyer accounts" do
+      buyer_without_applications = @provider.buyers.first
+
```
Comment (guicassolato) [inline test/integration/buyers/accounts_controller_test.rb:348]:
```suggestion
```

Considering https://github.com/3scale/porta/pull/2233/files#r515114897

**Review by guicassolato — APPROVED** [review]
LGTM. Nice work and thank you!

**Review by Martouta — CHANGES_REQUESTED** [review]
I have some proposals/comments to improve this, but I do not have time to test it and explain it properly now. I will try tonight or tomorrow morning :+1:

**Review by thomasmaas — DISMISSED** [review]

**Comment by HemaHG** [comment]
not a valid approach to display the accounts

**Pull request closed without merging**